frontend/src/relay/auth.tsx

```diff
@@ -0,0 +1,147 @@
+import { Link } from "../router";
+import { LinkButton } from "../ui/Button";
+
+
+/** Different external services we can link to using JWT pre-authentication */
+type JwtService = "EDITOR" | "STUDIO";
+
+const serviceUrl = (service: JwtService): URL => new URL(match(service, {
+    EDITOR: () => CONFIG.opencast.editorUrl,
+    STUDIO: () => CONFIG.opencast.studioUrl,
+}));
+
+type ExternalLinkProps = PropsWithChildren<{
+    className?: string;
+    fallback: "link" | "button";
+} & ({
+    service: "STUDIO";
+    params: {
+        "return.target": URL;
+    };
+} | {
+    service: "EDITOR";
+    params: {
+        mediaPackageId: string;
+    };
+})>;
+
+/**
+ * Authenticate a link to one of the Opencast hosted services we link to
+ * when `pre_auth_external_links` is enabled.
+ * In that case, this renders a form `POST`-ing you to Opencasts redirect endpoint
+ * passing along a JWT to get an authenticated session.
+ * The redirection then takes you to the final destination.
+ * If the setting is not enabled, this is just a normal link.
+ */
+export const ExternalLink: React.FC<ExternalLinkProps> = ({
+    className,
+    fallback,
+    service,
+    children,
+    params,
+}) => {
+    const target = serviceUrl(service);
+    for (const [key, value] of Object.entries(params)) {
+        target.searchParams.append(key, value);
+    }
+
+    if (!CONFIG.auth.preAuthExternalLinks) {
+        return match(fallback, {
+            link: () => <Link className={className} to={target.toString()} target="_blank">
+                {children}
+            </Link>,
+            button: () => <LinkButton className={className} to={target.toString()} target="_blank">
+                {children}
+            </LinkButton>,
+        });
+    }
+
+    const redirect = new URL("/redirect/get", target);
+
+    const onSubmit = async (event: FormEvent) => {
+        event.preventDefault();
+
+        const query = graphql`
+            query authLinkTokenQuery($service: JwtService!) {
+                jwt(service: $service)
+            }
+        `;
+        const result = await getJwt<authLinkTokenQuery>(query, { service });
+
+        const form = event.target as HTMLFormElement & { jwt: HTMLInputElement };
+        form.jwt.value = result.jwt;
+
+        form.submit();
+    };
+
+    return <form action={redirect.toString()} method="POST" target="_blank" onSubmit={onSubmit}>
+        <input type="hidden" name="target" value={target.toString()} />
+        <input type="hidden" name="jwt" value="test" />
+
+        <button className={className}>
+            {children}
+        </button>
+    </form>;
+};
+
+/** Fetches a new JWT for uploading to Opencast */
+export const getUploadJwt = async (): Promise<string> => {
+    const query = graphql`
+        query authUploadTokenQuery {
+            jwt(service: UPLOAD)
+        }
+    `;
+    const result = await getJwt<authUploadTokenQuery>(query);
+    return result.jwt;
+};
```

Comment:
Wouldn't it be easier to just completely reuse the `getJwt` here? Then `getJwt` does not need to be generic over the query parameter and just gets a single `UPLOAD | STUDIO | EDITOR` argument. The type of that is probably generated somewhere.

Reply:
Right, this went through a few iterations as well. At some point this wasn't possible, but now it is.